pkg/netpol/eval/internal/k8s/representative_selectors.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package k8s
+
+import (
+	"crypto/sha1" //nolint:gosec // Non-crypto use
+	"encoding/hex"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+// this file handles checking matching of representative peers' selectors' requirements
+// with rule selectors from policies
+// since a representative peer is a peer inferred from a network-policy rule
+
+// SelectorsFullMatch checks if there is a match between two input labelSelector objects,
+// one from a policy rule and the other from a representative pod.
+// for example, if a policy that captures a real workload `A` has a rule permitting egress connection,
+// with selectors that match a representative peer `B`, we need to capture this match,
+// and thus infer the exposure connectivity from `A` to `B`.
+// returns true if:
+// 1. both selectors point to same reference (rule and its matching representative pod/ns)
+// 2. if the rule's selector is empty (matches all pods/namespaces)
+// 3. if the requirements of both rules are equal (same)
+// i.e. checks if the selector from the policy rule fully matches the selector of the representative peer
+// note that : for non-empty selectors that partly match/ contained in each other the func returns false
+func SelectorsFullMatch(ruleSelector, repSelector *v1.LabelSelector) (bool, error) {
+	if ruleSelector == repSelector { // both label selectors point to same reference
+		return true, nil
+	}
+	// otherwise, check requirements equality
+	selector1, err := v1.LabelSelectorAsSelector(ruleSelector)
+	if err != nil {
+		return false, err
+	}
+	if selector1.Empty() { // empty rule matches everything
+		return true, nil
+	}
+	selector2, err := v1.LabelSelectorAsSelector(repSelector)
+	if err != nil {
+		return false, err
+	}
+	// Requirements() returns sorted by key list;
+	// since `v1.LabelSelectorAsSelector` builds a new `labels.selector` by using `Requirements.Add` which sorts the requirements slice by key.
+	// and `Requirements()` returns this sorted list
+	// links:
```

Comment:
rename requirements1,requirements2, to ruleRequirements and repRequirements

Reply:
done